lattices_macro/src/lib.rs

```diff
@@ -0,0 +1,560 @@
+        }
+    }
+}
+
+/// Renames the generics and returns the updated `WherePredicate`s.
+fn rename_generics(
+    item_struct: &mut ItemStruct,
+    rename: impl FnMut(&Ident) -> Ident,
+) -> Vec<WherePredicate> {
+    struct RenameGenerics<F> {
+        rename: F,
+        names: Vec<Ident>,
+        pub triggered: bool,
+    }
+    impl<F> VisitMut for RenameGenerics<F>
+    where
+        F: FnMut(&Ident) -> Ident,
+    {
+        fn visit_ident_mut(&mut self, i: &mut Ident) {
+            if self.names.contains(i) {
+                *i = (self.rename)(i);
+                self.triggered = true;
+            }
+        }
+    }
+
+    let names = item_struct
+        .generics
+        .type_params()
+        .map(|type_param| type_param.ident.clone())
+        .collect();
+    let mut visit = RenameGenerics {
+        rename,
+        names,
+        triggered: false,
+    };
+
+    let mut out = Vec::new();
+    if let Some(where_clause) = &mut item_struct.generics.where_clause {
+        for where_predicate in where_clause.predicates.iter_mut() {
+            visit.visit_where_predicate_mut(where_predicate);
+            if std::mem::take(&mut visit.triggered) {
+                out.push(where_predicate.clone());
+            }
+        }
+    }
+    for type_param in item_struct.generics.type_params_mut() {
+        visit.visit_type_param_mut(type_param);
+    }
+    for field in item_struct.fields.iter_mut() {
+        visit.visit_type_mut(&mut field.ty);
+    }
+    out
+}
+
+/// Ensures that `punctuated` has trailing punctuation (or is empty).
+fn ensure_trailing<T, P>(punctuated: &mut Punctuated<T, P>)
+where
+    P: Default,
+{
+    if !punctuated.empty_or_trailing() {
+        punctuated.push_punct(Default::default());
+    }
+}
+
+/// Derives `Merge`, `PartialEq`, `PartialOrd`, `LatticeOrd`, `IsBot`, `IsTop`, and `LatticeFrom`,
+/// and therefore `Lattice` too.
+///
+/// Can be thought of as shorthand for `#[derive(Merge, LatticeOrd, IsBot, IsTop, LatticeFrom)]`.
+///
+/// The lattice derived will be the _product lattice_ of all the fields of the struct. I.e a `Pair`
+/// lattice but of all the fields of the struct instead of just two.
+///
+/// These derive macros will create a second set of generics to allow conversion and merging
+/// between varying types. For example, given this struct:
+/// ```rust,ignore
+/// #[derive(Lattice)]
+/// struct MyLattice<KeySet, Epoch>
+/// where
+///     KeySet: Collection,
+///     Epoch: Ord,
+/// {
+///     keys: SetUnion<KeySet>,
+///     epoch: Max<Epoch>,
+/// }
+/// ```
+/// Will create derive macros `impl`s in the form:
+/// ```rust,ignore
+/// impl<KeySet, Epoch, KeySetOther, EpochOther>
+///     Merge<MyLattice<KeySetOther, EpochOther>>
+///     for MyLattice<KeySet, Epoch>
+/// where
+///     KeySet: Collection,
+///     Epoch: Ord,
+///     KeySetOther: Collection,
+///     EpochOther: Ord,
+/// {
+///     // ...
+/// }
+/// ```
+#[proc_macro_derive(Lattice)]
```

Comment:
Caution on naming: Using `Lattice` for the name reserves a fairly broad keyword for a specific use here. Consider if you want to use the name for something else later. The answer could be "no," which I am OK with.

Reply:
This is to match the `Lattices` alias in the lattices crate, so will be imported alongside that trait:
```rust
/// Alias trait for lattice types.
#[sealed]
pub trait Lattice: Sized + Merge<Self> + LatticeOrd + NaiveLatticeOrd + IsBot + IsTop {}
#[sealed]
impl<T> Lattice for T where T: Sized + Merge<Self> + LatticeOrd + NaiveLatticeOrd + IsBot + IsTop {}
```